### ISSUE TITLE
no-unresolved, check case even on case-insensitive file systems

### ISSUE THREAD

**Issue by mglidden** [benmosher/eslint-plugin-import#806]
Our team develops on Macs (case-insensitive file systems), but our build server is on a case-sensitive file system. This leads to occasional issues where `import Foo from 'Foo'` but the file was accidentally named `foo.js`. This will build on our development machines, but break on the build server. 

There's an option on [no-unresolved](https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-unresolved.md) to check case, but it only works on case-insensitive file systems.

Ideally, the case check would also work on a case-sensitive FS, or there would be another flag to enable it for a case-sensitive FS.

If this seems reasonable, I'm happy to work on a PR.

**Comment by ljharb** [comment]
I think maybe a third value to caseSensitive of "always", that enforced that the filename matched the code, would be great.

**Comment by johanlajili** [comment]
Note that it is a problem even on case insensitive file system (on babel + webpack that is). Sure, it will work, but it will imports the module as many time as there are written version. Especially bad for singletons.
fileA.js =>
```javascript 
export foo = {}
```
fileB.js =>
```javascript
import foo from 'fileA.js'
export const bar = foo;
```
fileC.js =>
```javascript
import foo from 'FileA.js' //note the wrong case used
import bar from 'fileB.js'
console.log(foo === bar) // false. Would be true if correct case was used.
```
This may sound like a weird edge case, but besides the obvious performance problem (including a same file multiple time is not exactly good), it can create weird bugs or invalidate test cases and it takes some times to realise the issue. eslint should be able to pick that up.

**Comment by benmosher** [comment]
I'm confused. The reason it's only implemented for case-insensitive filesystems is because on an FS where case _does_ matter, the file literally doesn't exist and thus is flagged as unresolved.

You're saying you'd like the reported message to be consistent across filesystems?

**Comment by johanlajili** [comment]
Using Webpack and Babel as the compiler, in a case-insensitive filesystem, the file is resolved regardless of the scale, **but** they are as many instances of the file as they are versions of the case written, which makes for some nasty bugs.
say you have **myFile.js** that does
````javascript
console.log('Hello World');
````
now if you import myFile in 50 different files like so:
````javascript
import 'myFile.js'
````
you will have only one console.log printed on screen. (expected behaviour).
But, if in one of those files you do instead
````javascript
import 'myfile.js'
````
You will now have two logs in the console. This proves particularly problematic if you export a singleton in a file, because if you get the case of the import wrong, you will get a different instance of the singleton.

If ESlint could enforce the file name to be correct with an "always" rule, that would be quite convenient.

**Comment by benmosher** [comment]
Sure, I get that, but that is what the `caseSensitive` option (on by default) already enforces: https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-unresolved.md#casesensitive

**Comment by benmosher** [comment]
See tests here: https://github.com/benmosher/eslint-plugin-import/blob/master/tests/src/rules/no-unresolved.js#L214

Again, tests aren't run / feature not implemented when filesystem is case-sensitive because the file technically doesn't exist if the case doesn't match, so the normal reporting will engage and report the import.

Closing because it's not clear what is missing.

**Comment by TimeBomb** [comment]
+1 to this.
We have developers on Linux and Mac OS.
Mac OS has a case insensitive file system.
Linux has a case sensitive filesystem.

We've had issues where a Mac developer committed code with a filename that did not match the exact case of the saved file. Mac OS doesn't care about this, so there are no linting errors. But this broke our code for Linux users, because it couldn't resolve the path due to incorrect casing.

It'd be great if this ESLint plugin enforced strict case sensitivity requirements on all filesystems, to ensure Linux users no longer have problems caused by Mac users failing to use the correct case sensitivity. I believe that's also what the original poster meant.

**Comment by charlessuh** [comment]
@benmosher: With Yarn 2 and PnP, the resolution of files in `node_modules` is handled by a virtual filesystem mechanism. See: https://yarnpkg.com/features/pnp#packages-are-stored-inside-zip-archives-how-can-i-access-their-files.

It appears this virtual mechanism is case sensitive while my filesystem on macOS (APFS) is not, so this rule incorrectly detects my filesystem as case sensitive and does not flag incorrect imports.

**Comment by ljharb** [comment]
@charlessuh pnp breaks a not insignificant amount of the ecosystem; you're welcome to file a new issue (this one is 3 years closed), but typically yarn 2, not every part of the ecosystem it breaks, needs to make a fix for things like this.
